ISSUER_PROTOCOL.md

```diff
@@ -0,0 +1,343 @@
+
+
+## Trust Token Crypto Protocol
+
+The Trust Token crypto protocol is based on the PMBTokens design in https://eprint.iacr.org/2020/072 (appendix H) using P-384. The necessary keys and function mappings are described below.
+
+
+### Keys
+
+The Trust Token protocol primarily requires two kinds of keypairs:
+
+*   SRRSigningKey/SRRVerificationKey - A Ed25519 public keypair for signing and verifying the integrity of the Signed Redemption Record response.
+*   TrustTokenSecretKey/TrustTokenPublicKey - A Trust Token keypair used to sign and verify Trust Tokens.
+
+#### SRRSigningKey/SRRVerificationKey
+
+These keys should be generated and stored as standard Ed25519 keys, with the public/verification key being included in the Key Commitment result as a 32-byte RFC8032 encoding.
+
+
+#### TrustTokenSecretKey/TrustTokenPublicKey
+
+These are keys used in Trust Token consisting of elliptic curve scalars and points. All scalars and points are sized based on the curve choice (P-384). Up to 3 keys may be configured in parallel in the key commitment.
+
+Each keypair consist of the following:
+
+
+```
+struct {
+  // Corresponding to the FALSE private metadata bit (x0, y0).
+  Scalar x0;
+  Scalar y0;
+  // Corresponding to the TRUE private metadata bit (x1, y1).
+  Scalar x1;
+  Scalar y1;
+  // Corresponding to the token validity (x˜, y˜).
+  Scalar xs;
+  Scalar ys;
+} TrustTokenSecretKey;
+
+struct {
+  ECPoint pub0; // Corresponding to the FALSE private metadata bit.
+  ECPoint pub1; // Corresponding to the TRUE private metadata bit.
+  ECPoint pubs; // Corresponding to the token validity check.
+} TrustTokenPublicKey;
+```
+
+The TrustTokenPublicKey is encoded in the key commitment and should be serialized as follows:
+
+
+```
+struct {
+  opaque pub0<1..2^16-1>; // X9.62 Uncompressed point.
+  opaque pub1<1..2^16-1>; // X9.62 Uncompressed point.
+  opaque pubs<1..2^16-1>; // X9.62 Uncompressed point.
+} TrustTokenPublicKey;
+```
+
+### Serialization/Hashing
+
+For the PMBTokens functions, the following serialization schemes and hashes are used internally using draft 07 of the hash-to-curve specification (https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07):
+
+
+```
+Ht(t) - P384_XMD:SHA-512_SSWU_RO_ with a big-endian bytestring input of 't' and a dst of "PMBTokens Experiment V1 HashT".
+
+Hs(T, s) - P384_XMD:SHA-512_SSWU_RO_ with the following input and a dst of "PMBTokens Experiment V1 HashS".
+
+struct {
+  opaque t<1..2^16-1>; // X9.62 Uncompressed point.
+  opaque s[Nn];
+};
+
+Hc(x) - P384_XMD:SHA-512_SSWU_RO_ scalar derivation (using the hash-to-curve hash_to_field operation modulo the group order rather than the field element) with the input x and a dst of "PMBTokens Experiment V1 HashC". Depending on the context, one of the following serializations is used for the input:
```

Comment:
Nit: group order and field element aren't analogous terms. It's extra fun since `hash_to_field` is apparently defined to allow an extension degree. Perhaps:

The hash-to-curve document does not define hash to scalars, so `Hc(x)` is defined to be the output of the [hash_to_field](https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-07#section-5.2) function with the following parameters:

* `DST` is "PMBTokens Experiment V1 HashC"
* `F`, `p`, and `m` are defined according to the finite field `GF(r)`, where `r` is the order of P-384. Note this is a different modulus from `hash_to_field` as used in P384_XMD:SHA-512_SSWU_RO_.
* `L` is 72, derived based on P384_XMD:SHA-512_SSWU_RO_'s security parameter `k` (192), and `p` defined above.
* `expand_message` uses the corresponding function from P384_XMD:SHA-512_SSWU_RO_.

When used in the DLEQ proof, `Hc` takes the following input.

...

When used in the DLEQOR proof, `Hc` takes the following input.

...

When used in the DLEQ batching step, `Hc` is called once for each `e_i` output, with the following input. The `index` field contains `i`.

...

(I wonder if we should just give them separate names. They're morally different hash functions.)

Reply:
Done, this will be cleaner once we get the hash to scalar.